Debugging medium example possibly hanging on Edison

Something on Edison changed, I cannot compile `toast` on Edison with the old `toast-deps`, a test recommended by @tskisner:

    module load toast-deps/20170707

See all modules:

```
module list
Currently Loaded Modulefiles:
  1) modules/3.2.10.6                              14) dmapp/7.1.1-6.0.4.0_46.2__gb8abda2.ari
  2) nsg/1.2.0                                     15) gni-headers/5.0.11-6.0.4.0_7.2__g7136988.ari
  3) gcc/6.3.0                                     16) xpmem/2.2.2-6.0.4.0_3.1__g43b0535.ari
  4) craype-ivybridge                              17) job/2.2.2-6.0.4.0_8.2__g3c644b5.ari
  5) craype-network-aries                          18) dvs/2.7_2.2.31-6.0.4.1_6.1__gb3b87e6
  6) craype/2.5.12                                 19) alps/6.4.1-6.0.4.0_7.2__g86d0f3d.ari
  7) cray-mpich/7.6.0                              20) rca/2.2.15-6.0.4.1_13.1__g46acb0f.ari
  8) altd/2.0                                      21) atp/2.1.1
  9) darshan/3.1.4                                 22) PrgEnv-gnu/6.0.4
 10) cray-libsci/17.06.1                           23) intel/17.0.2.174
 11) udreg/2.3.2-6.0.4.0_12.2__g2f9c3ee.ari        24) cmake/3.8.1
 12) ugni/6.0.14-6.0.4.0_14.1__ge7db4a2.ari        25) toast-deps/20170707
 13) pmi/5.0.12
```

Is it expected that I have cray, intel and gnu all loaded simultaneously?

On today's `toast` master:

    ./autogen.sh
    module load fftw # it looks like it needs fftw or MKL
    ./platforms/edison-gcc.sh --prefix=$PREFIX

I get this link error:

```
libtool: link: CC -DNDEBUG -fopenmp -O3 -g -fPIC -o .libs/toast_test toast_test.o -fopenmp  ./.libs/libtoast.so /opt/cray/pe/fftw/3.3.4.6/sandybridge/lib/libfftw3f_mpi.so /opt/cray/pe/fftw/3.3.4.6/sandybridge/lib/libfftw3f_threads.so /opt/cray/pe/fftw/3.3.4.6/sandybridge/lib/libfftw3f.so /opt/cray/pe/fftw/3.3.4.6/sandybridge/lib/libfftw3_mpi.so -lmpich /opt/gcc/6.3.0/snos/lib/../lib64/libgfortran.so /opt/gcc/6.3.0/snos/lib/../lib64/libquadmath.so /opt/gcc/6.3.0/snos/lib/../lib64/libstdc++.so -lEl /opt/cray/pe/fftw/3.3.4.6/sandybridge/lib/libfftw3_threads.so /opt/cray/pe/fftw/3.3.4.6/sandybridge/lib/libfftw3.so -lm -pthread -fopenmp -Wl,-rpath -Wl,/opt/gcc/6.3.0/snos/lib/../lib64 -Wl,-rpath -Wl,/opt/cray/pe/fftw/3.3.4.6/sandybridge/lib
./.libs/libtoast.so: undefined reference to `__intel_sse4_strncmp'
```

@tskisner I am not sure it is worth debugging this issue just for a test, what do you think?

If I try to build with intel, with the latest `toast-deps`, there is a configuration problem, possibly due to the fact that `cc` is pointing to Cray's compiler:

```
./platforms/edison-intel.sh --prefix=$PREFIX
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for style of include used by make... GNU
checking whether to compile using MPI... yes
checking for gcc... cc
checking whether the C compiler works... no
configure: error: in `/global/homes/z/zonca/p/software/toast':
configure: error: C compiler cannot create executables
See `config.log' for more details
edison[edison11] ~/p/software/toast master$ which cc
/opt/cray/pe/craype/2.5.12/bin/cc
```

Hi Andrea,

The toast-deps module loads the intel programming environment and overloads
the compiler wrappers. You can check like this:

cc -v
icc version 18.0.0 (gcc version 6.3.0 compatibility)
...

I just configured and installed the toast master branch on Edison without
an issue. Here is my shell function for the environment on Edison:

  edideps () {
      export PS1="\e[0;35m[edideps]\e[m $(ppwd \l)\h:\W> "
      module unload altd
      module use /global/common/software/cmb/edison/modulefiles
      module load toast-deps/20171126
      export
TOASTROOT=/global/common/software/cmb/edison/toast-deps/20171126_conda
      export
AUXROOT=/global/common/software/cmb/edison/toast-deps/20171126_aux
      export PREFIX=/scratch3/scratchdirs/keskital/software-toastdeps
      unset PYTHONSTARTUP
      export CRAYPE_LINK_TYPE=dynamic
      export PYTHONPATH=$PREFIX/lib/python3.6/site-packages:$PYTHONPATH
      export CC=icc
      export CXX=icpc
      export F77=ifort
      export FC=ifort
      export MPICC=cc
      export MPICXX=CC
      export OPENMP_CFLAGS="-qopenmp"
      export OPENMP_CXXFLAGS="-qopenmp"
      export CFLAGS="-O3 -g -fPIC -xavx -pthread"
      export CXXFLAGS="-O3 -g -fPIC -xavx -pthread"
      export FCFLAGS="-O3 -g -fPIC -xavx -pthread -heap-arrays 16"
      export CPPFLAGS="-I$PREFIX/include -I$PREFIX/include
-I${TOASTROOT}/include -I${AUXROOT}/include"
      export LDFLAGS="-qopenmp -L$PREFIX/lib -L${TOASTROOT}/lib
-L${AUXROOT}/lib"
      export PATH=$PREFIX/bin:$PATH
      export LD_LIBRARY_PATH=$PREFIX/lib:$LD_LIBRARY_PATH
      export LD_LIBRARY_PATH=$PREFIX/lib64:$LD_LIBRARY_PATH
      export LDSHARED="icc -shared"
  }

Not all of this is relevant for you but you might want to give it a go just
as it is.

-R



On Wed, Dec 6, 2017 at 5:21 AM, Andrea Zonca <notifications@github.com>
wrote:

> If I try to build with intel, with the latest toast-deps, there is a
> configuration problem, possibly due to the fact that cc is pointing to
> Cray's compiler:
>
> ./platforms/edison-intel.sh --prefix=$PREFIX
> checking for a BSD-compatible install... /usr/bin/install -c
> checking whether build environment is sane... yes
> checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
> checking for gawk... gawk
> checking whether make sets $(MAKE)... yes
> checking whether make supports nested variables... yes
> checking build system type... x86_64-unknown-linux-gnu
> checking host system type... x86_64-unknown-linux-gnu
> checking for style of include used by make... GNU
> checking whether to compile using MPI... yes
> checking for gcc... cc
> checking whether the C compiler works... no
> configure: error: in `/global/homes/z/zonca/p/software/toast':
> configure: error: C compiler cannot create executables
> See `config.log' for more details
> edison[edison11] ~/p/software/toast master$ which cc
> /opt/cray/pe/craype/2.5.12/bin/cc
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/hpc4cmb/toast/issues/153#issuecomment-349637631>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAkZGgeCVVxrKBtgiVQ0p_L2Wj-oQup7ks5s9pTLgaJpZM4Q331e>
> .
>

--- a/external/tools/modulefile.in
+++ b/external/tools/modulefile.in
@@ -16,9 +16,14 @@ proc ModulesHelp { } {
   puts stderr ""
 }
 
-conflict toastdeps
+conflict toast-deps
 
 @modload@
+
+# altd is evil and causes random job hangs
+if [ is-loaded altd ] {
+  module unload altd
+}
 
 # This is set by OS python and gives us problems.
 unsetenv PYTHONSTARTUP

--- a/pipelines/toast_satellite_sim.py
+++ b/pipelines/toast_satellite_sim.py
@@ -7,9 +7,11 @@
 from toast.mpi import MPI
 
 import os
+import sys
 
 import re
 import argparse
+import traceback
 
 import pickle
 
@@ -27,7 +29,8 @@ from toast.vis import set_backend
 def main():
 
     if MPI.COMM_WORLD.rank == 0:
-        print("Running with {} processes".format(MPI.COMM_WORLD.size))
+        print("Running with {} processes".format(MPI.COMM_WORLD.size),
+            flush=True)
 
     global_start = MPI.Wtime()
 
@@ -124,7 +127,7 @@ def main():
     if MPI.COMM_WORLD.size % groupsize != 0:
         if MPI.COMM_WORLD.rank == 0:
             print("WARNING:  process groupsize does not evenly divide into "
-                "total number of processes")
+                "total number of processes", flush=True)
     comm = toast.Comm(world=MPI.COMM_WORLD, groupsize=groupsize)
 
     # get options
@@ -167,7 +170,8 @@ def main():
     stop = MPI.Wtime()
     elapsed = stop - start
     if comm.comm_world.rank == 0:
-        print("Create focalplane:  {:.2f} seconds".format(stop-start))
+        print("Create focalplane:  {:.2f} seconds".format(stop-start), 
+            flush=True)
     start = stop
 
     if args.debug:
@@ -184,7 +188,8 @@ def main():
 
     if groupsize > len(fp.keys()):
         if comm.comm_world.rank == 0:
-            print("process group is too large for the number of detectors")
+            print("process group is too large for the number of detectors",
+                flush=True)
             comm.comm_world.Abort()
 
     # Detector information from the focalplane
@@ -264,7 +269,7 @@ def main():
     elapsed = stop - start
     if comm.comm_world.rank == 0:
         print("Read parameters, compute data distribution:  "
-            "{:.2f} seconds".format(stop-start))
+            "{:.2f} seconds".format(stop-start), flush=True)
     start = stop
 
     # we set the precession axis now, which will trigger calculation
@@ -286,7 +291,7 @@ def main():
     elapsed = stop - start
     if comm.comm_world.rank == 0:
         print("Construct boresight pointing:  "
-            "{:.2f} seconds".format(stop-start))
+            "{:.2f} seconds".format(stop-start), flush=True)
     start = stop
 
     # make a Healpix pointing matrix.
@@ -299,7 +304,7 @@ def main():
     stop = MPI.Wtime()
     elapsed = stop - start
     if comm.comm_world.rank == 0:
-        print("Pointing generation took {:.3f} s".format(elapsed))
+        print("Pointing generation took {:.3f} s".format(elapsed), flush=True)
     start = stop
 
     # Mapmaking.  For purposes of this simulation, we use detector noise
@@ -313,7 +318,7 @@ def main():
 
     if not args.madam:
         if comm.comm_world.rank == 0:
-            print("Not using Madam, will only make a binned map!")
+            print("Not using Madam, will only make a binned map!", flush=True)
 
         # get locally hit pixels
         lc = tm.OpLocalPixels()
@@ -349,7 +354,8 @@ def main():
         stop = MPI.Wtime()
         elapsed = stop - start
         if comm.comm_world.rank == 0:
-            print("Building hits and N_pp^-1 took {:.3f} s".format(elapsed))
+            print("Building hits and N_pp^-1 took {:.3f} s".format(elapsed),
+                flush=True)
         start = stop
 
         hits.write_healpix_fits("{}_hits.fits".format(args.outdir))
@@ -359,7 +365,8 @@ def main():
         stop = MPI.Wtime()
         elapsed = stop - start
         if comm.comm_world.rank == 0:
-            print("Writing hits and N_pp^-1 took {:.3f} s".format(elapsed))
+            print("Writing hits and N_pp^-1 took {:.3f} s".format(elapsed),
+                flush=True)
         start = stop
 
         # invert it
@@ -369,7 +376,8 @@ def main():
         stop = MPI.Wtime()
         elapsed = stop - start
         if comm.comm_world.rank == 0:
-            print("Inverting N_pp^-1 took {:.3f} s".format(elapsed))
+            print("Inverting N_pp^-1 took {:.3f} s".format(elapsed),
+                flush=True)
         start = stop
 
         invnpp.write_healpix_fits("{}_npp.fits".format(args.outdir))
@@ -378,7 +386,8 @@ def main():
         stop = MPI.Wtime()
         elapsed = stop - start
         if comm.comm_world.rank == 0:
-            print("Writing N_pp took {:.3f} s".format(elapsed))
+            print("Writing N_pp took {:.3f} s".format(elapsed),
+                flush=True)
         start = stop
 
         # in debug mode, print out data distribution information
@@ -395,7 +404,7 @@ def main():
             elapsed = stop - start
             if comm.comm_world.rank == 0:
                 print("Dumping debug data distribution took "
-                    "{:.3f} s".format(elapsed))
+                    "{:.3f} s".format(elapsed), flush=True)
             start = stop
 
         mcstart = start
@@ -417,7 +426,7 @@ def main():
             elapsed = stop - start
             if comm.comm_world.rank == 0:
                 print("Creating output dir {:04d} took {:.3f} s".format(mc,
-                    elapsed))
+                    elapsed), flush=True)
             start = stop
 
             # clear all noise data from the cache, so that we can generate
@@ -443,7 +452,7 @@ def main():
             elapsed = stop - start
             if comm.comm_world.rank == 0:
                 print("  Noise simulation {:04d} took {:.3f} s".format(mc,
-                    elapsed))
+                    elapsed), flush=True)
             start = stop
 
             zmap.data.fill(0.0)
@@ -457,7 +466,7 @@ def main():
             elapsed = stop - start
             if comm.comm_world.rank == 0:
                 print("  Building noise weighted map {:04d} took {:.3f} s".format(
-                    mc, elapsed))
+                    mc, elapsed), flush=True)
             start = stop
 
             tm.covariance_apply(invnpp, zmap)
@@ -467,7 +476,7 @@ def main():
             elapsed = stop - start
             if comm.comm_world.rank == 0:
                 print("  Computing binned map {:04d} took {:.3f} s".format(mc,
-                    elapsed))
+                    elapsed), flush=True)
             start = stop
 
             zmap.write_healpix_fits(os.path.join(outpath, "binned.fits"))
@@ -477,10 +486,11 @@ def main():
             elapsed = stop - start
             if comm.comm_world.rank == 0:
                 print("  Writing binned map {:04d} took {:.3f} s".format(mc,
-                    elapsed))
+                    elapsed), flush=True)
             elapsed = stop - mcstart
             if comm.comm_world.rank == 0:
-                print("  Mapmaking {:04d} took {:.3f} s".format(mc, elapsed))
+                print("  Mapmaking {:04d} took {:.3f} s".format(mc, elapsed),
+                    flush=True)
             start = stop
 
     else:
@@ -541,7 +551,8 @@ def main():
             stop = MPI.Wtime()
             elapsed = stop - start
             if comm.comm_world.rank == 0:
-                print("Noise simulation took {:.3f} s".format(elapsed))
+                print("Noise simulation took {:.3f} s".format(elapsed),
+                    flush=True)
             start = stop
 
             # create output directory for this realization
@@ -568,15 +579,23 @@ def main():
             stop = MPI.Wtime()
             elapsed = stop - start
             if comm.comm_world.rank == 0:
-                print("Mapmaking took {:.3f} s".format(elapsed))
+                print("Mapmaking took {:.3f} s".format(elapsed), flush=True)
 
     comm.comm_world.barrier()
     stop = MPI.Wtime()
     elapsed = stop - global_start
     if comm.comm_world.rank == 0:
-        print("Total Time:  {:.2f} seconds".format(elapsed))
+        print("Total Time:  {:.2f} seconds".format(elapsed), flush=True)
     MPI.Finalize()
 
 
 if __name__ == "__main__":
-    main()
+    try:
+        main()
+    except:
+        exc_type, exc_value, exc_traceback = sys.exc_info()
+        lines = traceback.format_exception(exc_type, exc_value, exc_traceback)
+        lines = [ "Proc {}: {}".format(MPI.COMM_WORLD.rank, x) for x in lines ]
+        print("".join(lines), flush=True)
+        MPI.COMM_WORLD.Abort()
+

